modules/gateway/nodes.go

```diff
@@ -90,68 +97,134 @@ func (g *Gateway) requestNodes(conn modules.PeerConn) error {
+					close(cancelDialChan)
+				}
+			}()
+			dialer := &net.Dialer{
+				Timeout: dialTimeout,
+				Cancel:  cancelDialChan,
+			}
+			conn, err := dialer.Dial("tcp", string(node))
 			if err != nil {
-				g.log.Debugf("WARN: RPC ShareNodes failed on peer %q: %v", peer, err)
-				continue
+				// NOTE: an error may be returned if the dial is cancelled
+				// partway through, which would cause the node to be pruned
+				// even though it may be a good node. Because nodes are
+				// plentiful, that's not a huge problem.
+				g.mu.Lock()
+				g.removeNode(node)
+				g.save()
+				g.mu.Unlock()
+				g.log.Debugf("INFO: removing node %q because dialing it failed: %v", node, err)
+				return
 			}
+
+			// If connection succeeds, supply an unacceptable version so that we
+			// will not be added as a peer.
+			//
+			// NOTE: this is a somewhat clunky way of specifying that you didn't
+			// actually want a connection.
+			encoding.WriteObject(conn, "0.0.0")
+			conn.Close()
+		}()
+	}
+}
+
+// permanentNodeManager tries to keep the Gateway's node list healthy. As long
+// as the Gateway has fewer than healthyNodeListLen nodes, it asks a random
+// peer for more nodes. It also continually pings nodes in order to establish
+// their connectivity. Unresponsive nodes are aggressively removed.
+func (g *Gateway) permanentNodeManager(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		// Wait 5 seconds so that a controlled number of node requests are made
+		// to peers.
+		select {
+		case <-time.After(nodeListDelay):
+		case <-g.threads.StopChan():
+			// Gateway is shutting down, close the thread.
+			return
 		}
 
-		// find an untested node to check
 		g.mu.RLock()
-		node, err := g.randomNode()
+		numNodes := len(g.nodes)
+		peer, err := g.randomPeer()
 		g.mu.RUnlock()
 		if err != nil {
+			// Most commonly the error indicates that there are no peers yet.
+			// Waiting through the next iteration gives the gateway time to
+			// bootstrap.
 			continue
 		}
 
-		// try to connect
-		conn, err := net.DialTimeout("tcp", string(node), dialTimeout)
-		if err != nil {
-			g.mu.Lock()
-			g.removeNode(node)
-			g.save()
-			g.mu.Unlock()
-			g.log.Debugf("INFO: removing node %q because dialing it failed: %v", node, err)
-			continue
-		}
-		// if connection succeeds, supply an unacceptable version to ensure
-		// they won't try to add us as a peer
-		encoding.WriteObject(conn, "0.0.0")
-		conn.Close()
-		// sleep for an extra 10 minutes after success; we don't want to spam
-		// connectable nodes
-		select {
-		case <-time.After(10 * time.Minute):
-		case <-g.threads.StopChan():
-			return
+		// Determine whether there are a satisfactory number of nodes in the
+		// nodelist. If there are not, use the random peer from earlier to
+		// expand the node list.
+		if numNodes < healthyNodeListLen {
+			// TODO: should this be done in a go func? It's a blocking
+			// function.
+			err := g.managedRPC(peer, "ShareNodes", g.requestNodes)
```

Comment:
The only `conn` we don't want to set a deadline on is the underlying connection to the peer. When we send or receive an RPC, a new "virtual" `conn` is created by muxado. We want to set a deadline on all of those. `grep` says we are not calling `SetDeadline` _anywhere_ in the gateway o.o

I'm surprised that the gateway is not already doing this, even on the receiving side. It would be really easy: just follow the host's example of setting a generous-but-finite deadline as soon as you accept the conn: https://github.com/NebulousLabs/Sia/blob/master/modules/gateway/rpc.go#L160
This may be a source of some gateway "deadlocks" or inefficiency.